tests/components/zeversolar_local/test_config_flow.py

```diff
@@ -0,0 +1,90 @@
+"""Test the Local access to the zeversolar invertor config flow."""
+from unittest.mock import patch
+
+import pytest
+from zeversolarlocal.api import ZeverError, default_url
+
+from homeassistant import config_entries
+from homeassistant.components.zeversolar_local import config_flow
+from homeassistant.components.zeversolar_local.const import DOMAIN, ZEVER_INVERTER_ID
+from homeassistant.const import CONF_HOST, CONF_URL
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import RESULT_TYPE_CREATE_ENTRY, RESULT_TYPE_FORM
+
+
+async def test_form(hass: HomeAssistant) -> None:
+    """Test we get the form."""
+
+    config_entry_data = {
+        "title": "Zeversolar invertor.",
+        ZEVER_INVERTER_ID: "inverterid",
+        CONF_URL: "http://0.0.0.0",
+    }
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == RESULT_TYPE_FORM
+    assert result["errors"] is None
+
+    with patch(
+        "homeassistant.components.zeversolar_local.config_flow.validate_input",
```

Comment:
Don't patch validate_input, but patch the library instead.

Reply:
I don't understand. Can you explain a bit more ?

Reply:
The validate_input function should not be tested separately. You need to patch the library [as you did in the validate_input test](https://github.com/home-assistant/core/pull/55440/files#diff-80be57af6c262a6f0205be7667b7d04c4f6ea3d2f8d3b73f6d412ea82bea1c94R81-R85) in the other tests too. So the test should not contain `patch("homeassistant.components.zeversolar_local.config_flow.a_config_flow_function"` but instead `patch("homeassistant.components.zeversolar_local.config_flow.library.a_library_function"`.

These are tests for the config flow, so we should not skip/patch any parts of the config_flow. We are not testing the library so we skip/patch the library.

Reply:
Ah. I understand. But I am not sure I agree. I have tested the `validate_input` already. So patching it while testing the actual `ConfigFlow` feels ok to me I guess.

Reply:
Or is my idea considered bad practice ?

Reply:
`validate_input` should not be tested in a separate test. :smile:
I think we don't allow to write a separate test for validate_input. Because like you said, it's bad practice.

The config flow test assumes that the given arguments and the return value of validate_input are correct.
The validate_input test assumes the config flow gives the correct arguments and assumes its return value is in the correct format for the config flow.

You fixed this by asserting the result of `validate_input` in the validate input test,
and setting it as the return value of the `validate_input`-patch in the config flow test.

So both functions are now tested separately but not together.
Example: If I change `info = await validate_input(self.hass, user_input)` to `info = await validate_input(self.hass)` all tests pass.

I can completely change the `validate_input` functions return value, arguments, etc and all the config flow test will pass.

The tests should be future-proof. They are not to test this PR but to prevent mistakes in future PRs.

Reply:
Ok. Makes sense. Thanks for the explanation! Learned something today !

Reply:
https://github.com/home-assistant/core/pull/55440/commits/66363ef86dfa7df1a6ef8161502d2678a77a1921